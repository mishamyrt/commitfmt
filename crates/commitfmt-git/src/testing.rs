use std::path::PathBuf;

use tempfile::TempDir;

use crate::command::run_git;
use crate::repository::Repository;
use crate::GitResult;

pub const COMMIT_HISTORY: &[&str] = &[
    "chore: initial commit",
    "feat(linter): remove description leading space rule",
    "feat(cc): trim description on parse",
    "fix: resolve parsing issue with special characters",
    "docs: update README with usage examples",
    "feat(cli): add new --verbose flag for detailed output",
    "test: add unit tests for commit message validation",
    "refactor: simplify configuration loading logic",
    "feat(format): implement auto-formatting for commit headers",
    "fix(git): handle empty repository edge case",
    "chore: update dependencies to latest versions",
    "feat: add support for custom commit templates",
    "fix(parser): correctly handle multiline commit bodies",
    "docs(api): add comprehensive API documentation",
    "feat(workspace): implement configuration inheritance",
    "test(integration): add end-to-end testing suite",
    "fix: prevent duplicate footers in formatted messages",
    "perf: optimize commit log parsing performance",
    "feat(hooks): add support for commit-msg hook",
    "ci: setup automated testing workflow",
];

pub struct TestBed {
    dir: TempDir,
    pub repo: Repository,
}

impl TestBed {
    pub const DEFAULT_BRANCH_NAME: &str = "main";
    pub const DEFAULT_USER_NAME: &str = "Test User";
    pub const DEFAULT_USER_EMAIL: &str = "test@example.com";

    /// Creates a new test bed with a fresh git repository
    pub fn empty() -> GitResult<Self> {
        let dir = TempDir::with_prefix("commitfmt-git-test-")?;
        let dir_path = dir.path();

        run_git(&["init"], dir_path)?;
        run_git(&["config", "user.name", Self::DEFAULT_USER_NAME], dir_path)?;
        run_git(&["config", "user.email", Self::DEFAULT_USER_EMAIL], dir_path)?;
        run_git(&["switch", "-c", Self::DEFAULT_BRANCH_NAME], dir_path)?;

        let repo = Repository::from_root(dir_path);

        Ok(Self { dir, repo })
    }

    /// Returns the path to the test bed
    pub fn path(&self) -> PathBuf {
        self.dir.path().to_path_buf()
    }

    /// Switches to a new branch
    pub fn switch_to_new(&self, branch_name: &str) -> GitResult<()> {
        run_git(&["switch", "-c", branch_name], self.dir.path())?;
        Ok(())
    }

    pub fn with_default_history() -> GitResult<Self> {
        let test_bed = Self::empty()?;
        for message in COMMIT_HISTORY {
            test_bed.repo.commit(message)?;
        }
        Ok(test_bed)
    }

    pub fn with_history(commits: &[&str]) -> GitResult<Self> {
        let test_bed = Self::empty()?;
        for message in commits {
            test_bed.repo.commit(message)?;
        }
        Ok(test_bed)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let test_bed = TestBed::empty().unwrap();
        assert!(test_bed.path().to_str().unwrap().contains("commitfmt-git-test-"));
        assert_eq!(test_bed.repo.get_branch_name(), Some("main".to_string()));
        assert_eq!(test_bed.repo.get_root(), test_bed.dir.path());
    }

    #[test]
    fn test_new_with_history() {
        let test_bed = TestBed::with_default_history().unwrap();
        assert_eq!(test_bed.repo.get_branch_name(), Some("main".to_string()));
        assert_eq!(test_bed.repo.get_root(), test_bed.dir.path());
        assert_eq!(
            test_bed.repo.run(&["rev-list", "--count", "--all"]).unwrap(),
            COMMIT_HISTORY.len().to_string()
        );
    }
}
