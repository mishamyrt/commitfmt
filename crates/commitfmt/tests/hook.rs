use std::fs::Permissions;

use commitfmt_git::testing::TestBed;
use insta::assert_snapshot;

fn write_hook(test_bed: &TestBed) {
    let exe: &str = env!("CARGO_BIN_EXE_commitfmt");
    let hook_path = test_bed.path().join(".git/hooks/prepare-commit-msg");

    std::fs::write(&hook_path, format!("#!/bin/sh\n\n{exe}")).unwrap();

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&hook_path, Permissions::from_mode(0o755)).unwrap();
    }
}

#[test]
fn test_hook_default() {
    let test_bed = TestBed::with_default_history().unwrap();
    let repo = test_bed.repo.clone();

    write_hook(&test_bed);

    repo.commit("feat(   test   ) : test").unwrap();

    let log = repo.get_log("HEAD~1", "HEAD").unwrap();
    assert_eq!(log[0].message, "feat(test): test\n");
}

#[test]
fn test_hook_with_config_correct_message() {
    let test_bed = TestBed::with_default_history().unwrap();
    let repo = test_bed.repo.clone();
    let config = r#"
[lint.header]
scope-enum = ["api", "core"]
type-enum = ["feat", "fix"]

[lint.footer]
exists = ["Issue-ID"]
"#;
    std::fs::write(test_bed.path().join(".commitfmt.toml"), config).unwrap();
    write_hook(&test_bed);

    repo.commit("feat (  api   ) : test\nbody\n\nIssue-ID: 123").unwrap();

    let log = repo.get_log("HEAD~1", "HEAD").unwrap();
    assert_eq!(log[0].message, "feat(api): test\n\nbody\n\nIssue-ID: 123\n");
}

#[test]
fn test_hook_with_config_incorrect_message() {
    let test_bed = TestBed::with_default_history().unwrap();
    let repo = test_bed.repo.clone();
    let config = r#"
[lint.header]
type-min-length = 4
max-length = 30
scope-max-length = 10

[lint.body]
min-length = 10

[lint.footer]
max-line-length = 10
"#;

    std::fs::write(test_bed.path().join(".commitfmt.toml"), config).unwrap();
    write_hook(&test_bed);

    let err = repo
        .commit(
            "fix(application): test message about feature

body

BREAKING-CHANGE: Description about my feature and a lot of text",
        )
        .unwrap_err();

    match err {
        commitfmt_git::GitError::CommandFailed(_, message) => {
            assert_snapshot!("error_with_config", &message);
        }
        _ => panic!("Expected GitError::CommandFailed"),
    }
}

#[test]
fn test_hook_append_footers() {
    let test_bed = TestBed::with_default_history().unwrap();
    let repo = test_bed.repo.clone();
    test_bed.switch_to_new("feature/CFMT-123/description").unwrap();

    let config = r#"
[[additional-footers]]
key = "Authored-by"
value-template = "{{ echo 'John Doe' }}"

[[additional-footers]]
key = "Ticket-ID"
branch-value-pattern = "(?:.*)/([A-Z0-9-]+)/?(?:.*)"
"#;
    std::fs::write(test_bed.path().join(".commitfmt.toml"), config).unwrap();

    write_hook(&test_bed);

    repo.commit("feat(test): test\nbody").unwrap();

    let log = repo.get_log("HEAD~1", "HEAD").unwrap();
    assert_eq!(
        log[0].message,
        "feat(test): test\n\nbody\n\nAuthored-by: John Doe\nTicket-ID: CFMT-123\n"
    );
}
