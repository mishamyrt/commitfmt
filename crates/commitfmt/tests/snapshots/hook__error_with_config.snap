---
source: crates/commitfmt/tests/hook.rs
assertion_line: 79
expression: "&message"
---
- Type is shorter than 4 characters [type-min-length]
- Scope is longer than 10 characters [scope-max-length]
- Header is longer than 30 characters [max-length]
- Body is shorter than 10 characters [min-length]
- Footer 'BREAKING-CHANGE' contains a line that length is longer than 10 characters [max-line-length]

Message has 5 problems
